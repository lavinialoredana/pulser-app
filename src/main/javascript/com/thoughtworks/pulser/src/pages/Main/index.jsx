import "././Main.css";

function Main() {
  return (
    <div className="Main">
      <header className="Main-header">
        <h1 data-testid="main-header"> How are you feeling today?  </h1>
      </header>
    </div>
  );
}

export default Main;