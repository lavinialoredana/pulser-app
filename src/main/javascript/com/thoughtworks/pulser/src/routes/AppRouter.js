import { createBrowserRouter } from "react-router-dom";
import ErrorPage from "../pages/Error";
import Main from "../pages/Main";
import PulserFeed from "../pages/PulserFeed";

export const routesConfig = [
  {
    path: "/",
    element: <Main />,
    errorElement: <ErrorPage />,
  },

  {
    path: "/pulserfeed/messages",
    element: <PulserFeed />,
    errorElement: <ErrorPage />,
  },
];

const AppRouter = createBrowserRouter(routesConfig);

export default AppRouter;
