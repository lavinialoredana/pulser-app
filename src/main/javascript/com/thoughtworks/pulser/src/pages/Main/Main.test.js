import { cleanup, render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { BrowserRouter, RouterProvider, Router } from "react-router-dom";
import Main from "./index.jsx";
import { routesConfig } from "../../routes/AppRouter.js";
import * as router from "react-router";
import { createMemoryHistory } from "@remix-run/router";

describe("Main Page structure", () => {
  afterEach(() => {
    cleanup();
  });

  it("should match snapshot", () => {
    const { asFragment } = render(<Main />, { wrapper: BrowserRouter });

    expect(asFragment()).toMatchSnapshot();
  });

  it("should render essential components", () => {
    render(<Main />, { wrapper: BrowserRouter });

    const mainHeaderElement = screen.getByTestId("main-header");
    const feedbackMessageComponent = screen.getByTestId("textarea-field");
    const reactionsComponent = screen.getByTestId("reactions");
    const feedbackReactionComponent =
      screen.getAllByTestId("feedback-reaction");
    const buttonComponent = screen.getByTestId("button");

    expect(mainHeaderElement).toBeInTheDocument();
    expect(feedbackMessageComponent).toBeInTheDocument();
    expect(reactionsComponent).toBeInTheDocument();
    expect(feedbackReactionComponent.length).toBe(4);
    expect(buttonComponent).toBeInTheDocument();
  });

  it("should render the title", () => {
    render(<Main />, { wrapper: BrowserRouter });

    const headerText = screen.getByText("How are you feeling today?");

    expect(headerText).toBeInTheDocument();
  });
});

describe("Main Page full user flow", () => {
  afterEach(() => {
    jest.restoreAllMocks();
    cleanup();
  });

  it("should match snapshot", () => {
    const { asFragment } = render(<Main />, { wrapper: BrowserRouter });

    expect(asFragment()).toMatchSnapshot();
  });

  it("should send", async () => {
    const user = userEvent.setup();
    const mockedFetch = jest.fn(() => {
      return Promise.resolve({
        status: 200,
        json: () => Promise.resolve({ success: "OK" }),
      });
    });

    render(<Main />, { wrapper: BrowserRouter });

    const faceKey = "AWESOME";
    const awesomeImage = screen.getByAltText("awesome_face");
    await user.click(awesomeImage);

    const message = "Thank you!";
    const textArea = screen.getByTestId("textarea-field");
    await user.type(textArea, message);

    jest.spyOn(global, "fetch").mockImplementation(mockedFetch);

    const submitButton = screen.getByTestId("button");
    await user.click(submitButton);

    expect(mockedFetch).toHaveBeenCalledTimes(1);
    expect(mockedFetch).toHaveBeenCalledWith(
      "https://reqbin.com/echo/post/json",
      {
        method: "POST",
        headers: {
          Accept: "application/json",
        },
        body: JSON.stringify({
          face: faceKey,
          body: message,
        }),
      }
    );
  });
});

describe("Router error page testing", () => {
  afterEach(() => {
    cleanup();
  });

  it("should verify page content for expected route after navigating", async () => {
    const user = userEvent.setup();
    const mockedFetch = jest.fn(() => {
      return Promise.resolve({
        status: 200,
        json: () => Promise.resolve({ success: "OK" }),
      });
    });

    const landingRoute = "/";
    const pulserFeedRoute = "/pulserfeed/messages";

    const history = createMemoryHistory(routesConfig, {
      initialEntries: [landingRoute],
    });

    render(
      <Router location={history} navigator={history}>
        <Main />
      </Router>
    );

    const awesomeImage = screen.getByAltText("awesome_face");
    await user.click(awesomeImage);

    const message = "Thank you!";
    const textArea = screen.getByTestId("textarea-field");
    await user.type(textArea, message);

    jest.spyOn(global, "fetch").mockImplementation(mockedFetch);

    expect(history.location.pathname).toBe(landingRoute);

    const submitButton = screen.getByTestId("button");
    await user.click(submitButton);

    expect(history.location.pathname).toBe(pulserFeedRoute);
  });

  test("landing on the main page", () => {
    const landingRoute = "/";
    const badRoute = "/some/bad/route";

    const history = createMemoryHistory(routesConfig, {
      initialEntries: [landingRoute],
    });

    render(
      <Router location={history} navigator={history}>
        <Main />
      </Router>
    );

    expect(history.location.pathname).not.toBe(badRoute);
    expect(history.location.pathname).toBe(landingRoute);
  });

  test("landing on the error page", () => {
    const badRoute = "/some/bad/route";

    const customMemoryRouter = router.createMemoryRouter(routesConfig, {
      initialEntries: [badRoute],
    });

    render(<RouterProvider router={customMemoryRouter} />);

    const errorPageTitleElement = screen.getByTestId("error-title");

    expect(customMemoryRouter.state.location.pathname).toBe(badRoute);
    expect(errorPageTitleElement).toBeInTheDocument();
  });
});

const navigateMock = jest.fn();

describe("Router happy path testing", () => {
  beforeEach(() => {
    jest.spyOn(router, "useNavigate").mockImplementation(() => navigateMock);
  });
  afterEach(() => {
    cleanup();
  });

  it("should call navigate on submit button", async () => {
    const user = userEvent.setup();
    const mockedFetch = jest.fn(() => {
      return Promise.resolve({
        status: 200,
        json: () => Promise.resolve({ success: "OK" }),
      });
    });

    render(<Main />, { wrapper: BrowserRouter });

    const awesomeImage = screen.getByAltText("awesome_face");
    await user.click(awesomeImage);

    const message = "Thank you!";
    const textArea = screen.getByTestId("textarea-field");
    await user.type(textArea, message);

    jest.spyOn(global, "fetch").mockImplementation(mockedFetch);

    const mockedPath = "/pulserfeed/messages";

    const submitButton = screen.getByTestId("button");
    await user.click(submitButton);

    expect(navigateMock).toHaveBeenCalledWith(mockedPath);
  });
});
