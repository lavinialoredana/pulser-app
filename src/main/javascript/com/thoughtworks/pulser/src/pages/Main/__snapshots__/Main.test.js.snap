// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main Page full user flow should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="Main"
    data-testid="main-page"
  >
    <header
      class="Main-header"
    >
      <h1
        data-testid="main-header"
      >
         How are you feeling today? 
      </h1>
    </header>
    <div>
      <div
        class="Reactions-container"
        data-testid="reactions"
      >
        <div>
          <div
            class="Feedback-reaction-component  "
            data-testid="feedback-reaction"
          >
            <h3
              data-testid="feedBackReaction-title"
            >
               Awesome
            </h3>
            <img
              alt="awesome_face"
              class="Feedback-reaction-image"
              src="awesome_face.png"
            />
          </div>
        </div>
        <div>
          <div
            class="Feedback-reaction-component  "
            data-testid="feedback-reaction"
          >
            <h3
              data-testid="feedBackReaction-title"
            >
               Happy
            </h3>
            <img
              alt="happy_face"
              class="Feedback-reaction-image"
              src="happy_face.png"
            />
          </div>
        </div>
        <div>
          <div
            class="Feedback-reaction-component  "
            data-testid="feedback-reaction"
          >
            <h3
              data-testid="feedBackReaction-title"
            >
               Sad
            </h3>
            <img
              alt="sad_face"
              class="Feedback-reaction-image"
              src="sad_face.png"
            />
          </div>
        </div>
        <div>
          <div
            class="Feedback-reaction-component  "
            data-testid="feedback-reaction"
          >
            <h3
              data-testid="feedBackReaction-title"
            >
               Angry
            </h3>
            <img
              alt="angry_face"
              class="Feedback-reaction-image"
              src="angry_face.png"
            />
          </div>
        </div>
      </div>
      <div>
        <textarea
          cols="80"
          data-testid="textarea-field"
          placeholder="Write here any additional feedback"
          rows="4"
        />
      </div>
      <div>
        <button
          data-testid="button"
          disabled=""
          name="Submit"
        >
          Submit
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Main Page structure should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="Main"
    data-testid="main-page"
  >
    <header
      class="Main-header"
    >
      <h1
        data-testid="main-header"
      >
         How are you feeling today? 
      </h1>
    </header>
    <div>
      <div
        class="Reactions-container"
        data-testid="reactions"
      >
        <div>
          <div
            class="Feedback-reaction-component  "
            data-testid="feedback-reaction"
          >
            <h3
              data-testid="feedBackReaction-title"
            >
               Awesome
            </h3>
            <img
              alt="awesome_face"
              class="Feedback-reaction-image"
              src="awesome_face.png"
            />
          </div>
        </div>
        <div>
          <div
            class="Feedback-reaction-component  "
            data-testid="feedback-reaction"
          >
            <h3
              data-testid="feedBackReaction-title"
            >
               Happy
            </h3>
            <img
              alt="happy_face"
              class="Feedback-reaction-image"
              src="happy_face.png"
            />
          </div>
        </div>
        <div>
          <div
            class="Feedback-reaction-component  "
            data-testid="feedback-reaction"
          >
            <h3
              data-testid="feedBackReaction-title"
            >
               Sad
            </h3>
            <img
              alt="sad_face"
              class="Feedback-reaction-image"
              src="sad_face.png"
            />
          </div>
        </div>
        <div>
          <div
            class="Feedback-reaction-component  "
            data-testid="feedback-reaction"
          >
            <h3
              data-testid="feedBackReaction-title"
            >
               Angry
            </h3>
            <img
              alt="angry_face"
              class="Feedback-reaction-image"
              src="angry_face.png"
            />
          </div>
        </div>
      </div>
      <div>
        <textarea
          cols="80"
          data-testid="textarea-field"
          placeholder="Write here any additional feedback"
          rows="4"
        />
      </div>
      <div>
        <button
          data-testid="button"
          disabled=""
          name="Submit"
        >
          Submit
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Router testing should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="Main"
    data-testid="main-page"
  >
    <header
      class="Main-header"
    >
      <h1
        data-testid="main-header"
      >
         How are you feeling today? 
      </h1>
    </header>
    <div>
      <div
        class="Reactions-container"
        data-testid="reactions"
      >
        <div>
          <div
            class="Feedback-reaction-component  "
            data-testid="feedback-reaction"
          >
            <h3
              data-testid="feedBackReaction-title"
            >
               Awesome
            </h3>
            <img
              alt="awesome_face"
              class="Feedback-reaction-image"
              src="awesome_face.png"
            />
          </div>
        </div>
        <div>
          <div
            class="Feedback-reaction-component  "
            data-testid="feedback-reaction"
          >
            <h3
              data-testid="feedBackReaction-title"
            >
               Happy
            </h3>
            <img
              alt="happy_face"
              class="Feedback-reaction-image"
              src="happy_face.png"
            />
          </div>
        </div>
        <div>
          <div
            class="Feedback-reaction-component  "
            data-testid="feedback-reaction"
          >
            <h3
              data-testid="feedBackReaction-title"
            >
               Sad
            </h3>
            <img
              alt="sad_face"
              class="Feedback-reaction-image"
              src="sad_face.png"
            />
          </div>
        </div>
        <div>
          <div
            class="Feedback-reaction-component  "
            data-testid="feedback-reaction"
          >
            <h3
              data-testid="feedBackReaction-title"
            >
               Angry
            </h3>
            <img
              alt="angry_face"
              class="Feedback-reaction-image"
              src="angry_face.png"
            />
          </div>
        </div>
      </div>
      <div>
        <textarea
          cols="80"
          data-testid="textarea-field"
          placeholder="Write here any additional feedback"
          rows="4"
        />
      </div>
      <div>
        <button
          data-testid="button"
          disabled=""
          name="Submit"
        >
          Submit
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`WIP ROUTER TESTING should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="Main"
    data-testid="main-page"
  >
    <header
      class="Main-header"
    >
      <h1
        data-testid="main-header"
      >
         How are you feeling today? 
      </h1>
    </header>
    <div>
      <div
        class="Reactions-container"
        data-testid="reactions"
      >
        <div>
          <div
            class="Feedback-reaction-component  "
            data-testid="feedback-reaction"
          >
            <h3
              data-testid="feedBackReaction-title"
            >
               Awesome
            </h3>
            <img
              alt="awesome_face"
              class="Feedback-reaction-image"
              src="awesome_face.png"
            />
          </div>
        </div>
        <div>
          <div
            class="Feedback-reaction-component  "
            data-testid="feedback-reaction"
          >
            <h3
              data-testid="feedBackReaction-title"
            >
               Happy
            </h3>
            <img
              alt="happy_face"
              class="Feedback-reaction-image"
              src="happy_face.png"
            />
          </div>
        </div>
        <div>
          <div
            class="Feedback-reaction-component  "
            data-testid="feedback-reaction"
          >
            <h3
              data-testid="feedBackReaction-title"
            >
               Sad
            </h3>
            <img
              alt="sad_face"
              class="Feedback-reaction-image"
              src="sad_face.png"
            />
          </div>
        </div>
        <div>
          <div
            class="Feedback-reaction-component  "
            data-testid="feedback-reaction"
          >
            <h3
              data-testid="feedBackReaction-title"
            >
               Angry
            </h3>
            <img
              alt="angry_face"
              class="Feedback-reaction-image"
              src="angry_face.png"
            />
          </div>
        </div>
      </div>
      <div>
        <textarea
          cols="80"
          data-testid="textarea-field"
          placeholder="Write here any additional feedback"
          rows="4"
        />
      </div>
      <div>
        <button
          data-testid="button"
          disabled=""
          name="Submit"
        >
          Submit
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
